File: src/core/audit_log.ts
import crypto from "crypto";
import Database from "better-sqlite3";

export enum AuditEventKind {
  SESSION_CREATED = "session_created",
  SESSION_CLOSED = "session_closed",
  TASK_CREATED = "task_created",
  TASK_COMPLETED = "task_completed",
  TASK_FAILED = "task_failed",
  TOOL_INVOKED = "tool_invoked",
  TOOL_ALLOWED = "tool_allowed",
  TOOL_DENIED = "tool_denied",
  APPROVAL_REQUESTED = "approval_requested",
  APPROVAL_GRANTED = "approval_granted",
  APPROVAL_DENIED = "approval_denied",
  PLUGIN_REGISTERED = "plugin_registered",
  PLUGIN_UNREGISTERED = "plugin_unregistered",
  CAPABILITY_TOKEN_ISSUED = "capability_token_issued",
  SANDBOX_EXECUTION = "sandbox_execution",
  HOST_ELEVATION = "host_elevation",
}

export interface AuditEvent {
  id: string;
  kind: AuditEventKind;
  actorId: string;
  timestamp: string;
  details: Record<string, unknown>;
}

interface AuditRow {
  id: string;
  kind: string;
  actor_id: string;
  timestamp: string;
  details: string;
}

export class AuditLog {
  private db: Database.Database;

  constructor(dbPath = ":memory:") {
    this.db = new Database(dbPath);
    this.db.exec(`
      CREATE TABLE IF NOT EXISTS audit_events (
        id TEXT PRIMARY KEY,
        kind TEXT NOT NULL,
        actor_id TEXT NOT NULL,
        timestamp TEXT NOT NULL,
        details TEXT NOT NULL
      )
    `);
  }

  record(kind: AuditEventKind, actorId: string, details: Record<string, unknown> = {}): AuditEvent {
    const event: AuditEvent = {
      id: crypto.randomUUID(),
      kind,
      actorId,
      timestamp: new Date().toISOString(),
      details,
    };
    this.db
      .prepare(
        "INSERT INTO audit_events (id, kind, actor_id, timestamp, details) VALUES (?, ?, ?, ?, ?)"
      )
      .run(event.id, event.kind, event.actorId, event.timestamp, JSON.stringify(event.details));
    return event;
  }

  events(): AuditEvent[] {
    return (this.db.prepare("SELECT * FROM audit_events ORDER BY timestamp ASC").all() as AuditRow[]).map(
      this.deserialize
    );
  }

  eventsByKind(kind: AuditEventKind): AuditEvent[] {
    return (
      this.db
        .prepare("SELECT * FROM audit_events WHERE kind = ? ORDER BY timestamp ASC")
        .all(kind) as AuditRow[]
    ).map(this.deserialize);
  }

  eventsByActor(actorId: string): AuditEvent[] {
    return (
      this.db
        .prepare("SELECT * FROM audit_events WHERE actor_id = ? ORDER BY timestamp ASC")
        .all(actorId) as AuditRow[]
    ).map(this.deserialize);
  }

  count(): number {
    const row = this.db.prepare("SELECT COUNT(*) as cnt FROM audit_events").get() as { cnt: number };
    return row.cnt;
  }

  close(): void {
    this.db.close();
  }

  private deserialize(row: AuditRow): AuditEvent {
    return {
      id: row.id,
      kind: row.kind as AuditEventKind,
      actorId: row.actor_id,
      timestamp: row.timestamp,
      details: JSON.parse(row.details) as Record<string, unknown>,
    };
  }
}

File: openclaw_safe/kernel/audit_log.py
"""AuditLog – append-only record of all significant kernel events."""

from __future__ import annotations

import uuid
from dataclasses import dataclass, field
from datetime import datetime, timezone
from enum import Enum
from typing import Any


class AuditEventKind(str, Enum):
    SESSION_CREATED = "session_created"
    SESSION_CLOSED = "session_closed"
    TASK_CREATED = "task_created"
    TASK_COMPLETED = "task_completed"
    TASK_FAILED = "task_failed"
    TOOL_INVOKED = "tool_invoked"
    TOOL_ALLOWED = "tool_allowed"
    TOOL_DENIED = "tool_denied"
    APPROVAL_REQUESTED = "approval_requested"
    APPROVAL_GRANTED = "approval_granted"
    APPROVAL_DENIED = "approval_denied"
    PLUGIN_REGISTERED = "plugin_registered"
    PLUGIN_UNREGISTERED = "plugin_unregistered"
    CAPABILITY_TOKEN_ISSUED = "capability_token_issued"
    SANDBOX_EXECUTION = "sandbox_execution"
    HOST_ELEVATION = "host_elevation"


@dataclass
class AuditEvent:
    kind: AuditEventKind
    actor_id: str
    details: dict[str, Any] = field(default_factory=dict)
    id: str = field(default_factory=lambda: str(uuid.uuid4()))
    timestamp: datetime = field(default_factory=lambda: datetime.now(timezone.utc))

    def __str__(self) -> str:
        return (
            f"AuditEvent({self.kind.value}, actor={self.actor_id!r}, "
            f"at={self.timestamp.isoformat()})"
        )


class AuditLog:
    """Thread-safe append-only audit log."""

    def __init__(self) -> None:
        self._events: list[AuditEvent] = []

    def record(
        self,
        kind: AuditEventKind,
        actor_id: str,
        **details: Any,
    ) -> AuditEvent:
        event = AuditEvent(kind=kind, actor_id=actor_id, details=dict(details))
        self._events.append(event)
        return event

    @property
    def events(self) -> list[AuditEvent]:
        return list(self._events)

    def events_by_kind(self, kind: AuditEventKind) -> list[AuditEvent]:
        return [e for e in self._events if e.kind == kind]

    def events_by_actor(self, actor_id: str) -> list[AuditEvent]:
        return [e for e in self._events if e.actor_id == actor_id]

    def __len__(self) -> int:
        return len(self._events)
